src/kv/MemDBStore.cc

```diff
@@ -0,0 +1,523 @@
+{
+  merge_ops.push_back(std::make_pair(prefix, mop));
+  return 0;
+}
+
+int MSStore::do_open(ostream &out, bool create)
+{
+  m_total_bytes = 0;
+  m_allocated_bytes = 0;
+
+  return _init(create);
+}
+
+MSStore::~MSStore()
+{
+  close();
+}
+
+void MSStore::close()
+{
+  /*
+   * Save whatever in memory btree.
+   */
+  _save();
+}
+
+int MSStore::submit_transaction(KeyValueDB::Transaction t)
+{
+  MSTransactionImpl* mt =  static_cast<MSTransactionImpl*>(t.get());
+
+  dtrace << __func__ << " " << mt->get_ops().size() << dendl;
+  for(auto& op : mt->get_ops()) {
+    if(op.first == MSTransactionImpl::WRITE) {
+      ms_op_t set_op = op.second;
+      _setkey(set_op);
+    } else if (op.first == MSTransactionImpl::MERGE) {
+      ms_op_t merge_op = op.second;
+      _merge(merge_op);
+    } else {
+      ms_op_t rm_op = op.second;
+      assert(op.first == MSTransactionImpl::DELETE);
+      _rmkey(rm_op);
+    }
+  }
+
+  return 0;
+}
+
+int MSStore::submit_transaction_sync(KeyValueDB::Transaction tsync)
+{
+  dtrace << __func__ << " " << dendl;
+  submit_transaction(tsync);
+  return 0;
+}
+
+int MSStore::transaction_rollback(KeyValueDB::Transaction t)
+{
+  MSTransactionImpl* mt =  static_cast<MSTransactionImpl*>(t.get());
+  mt->clear();
+  return 0;
+}
+
+void MSStore::MSTransactionImpl::set(
+  const string &prefix, const string &k, const bufferlist &to_set_bl)
+{
+  dtrace << __func__ << " " << prefix << " " << k << dendl;
+  ops.push_back(make_pair(WRITE, std::make_pair(std::make_pair(prefix, k),
+                  to_set_bl)));
+}
+
+void MSStore::MSTransactionImpl::rmkey(const string &prefix,
+    const string &k)
+{
+  dtrace << __func__ << " " << prefix << " " << k << dendl;
+  ops.push_back(make_pair(DELETE,
+                          std::make_pair(std::make_pair(prefix, k),
+                          NULL)));
+}
+
+void MSStore::MSTransactionImpl::rmkeys_by_prefix(const string &prefix)
+{
+  KeyValueDB::Iterator it = m_db->get_iterator(prefix);
+  for (it->seek_to_first(); it->valid(); it->next()) {
+    rmkey(prefix, it->key());
+  }
+}
+
+void MSStore::MSTransactionImpl::merge(
+  const std::string &prefix, const std::string &key, const bufferlist  &value)
+{
+
+  dtrace << __func__ << " " << prefix << " " << key << dendl;
+  ops.push_back(make_pair(MERGE, make_pair(std::make_pair(prefix, key), value)));
+  return;
+}
+
+int MSStore::_setkey(ms_op_t &op)
+{
+  std::lock_guard<std::mutex> l(m_lock);
+  std::string key = combine_strings(op.first.first, op.first.second);
+  bufferlist bl = op.second;
+
+  m_total_bytes += bl.length();
+
+  bufferlist bl_old;
+  if (_get(op.first.first, op.first.second, &bl_old)) {
+    /*
+     * delete and free existing key.
+     */
+    m_total_bytes -= bl_old.length();
+    m_btree.erase(key);
+  }
+
+  m_btree[key] = bufferptr((char *) bl.c_str(), bl.length());
+
+  return 0;
+}
+
+int MSStore::_rmkey(ms_op_t &op)
+{
+  std::lock_guard<std::mutex> l(m_lock);
```

Comment:
Yes, I will take utility code out of lock and break code wherever possible. Also, I have in mind overall concurrency improvement for this code using multiple btrees.